<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">

<head>

<meta charset="utf-8" />
<meta name="generator" content="quarto-1.3.433" />

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />


<title>Network Analysis in Systems Biology with R/Bioconductor – 2  Dealing with multiple data sets: consensus modules and module preservation</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<!-- htmldependencies:E3FAD763 -->
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body>

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn"
      data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" 
      aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation"
      onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <h1 class="quarto-secondary-nav-title"></h1>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" 
      aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation"
      onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="/">
      Network Analysis in Systems Biology with R/Bioconductor
      </a> 
        <div class="sidebar-tools-main">
    <a href="/Network-Analysis-in-Systems-Biology-with-R-Bioconductor
.pdf" rel="" title="Download PDF" class="quarto-navigation-tool px-1" aria-label="Download PDF"><i class="bi bi-file-pdf"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/chapters/00_setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Setup</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/chapters/01_inference_and_analysis_of_GCNs.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>1</span>  <span class='chapter-title'>Inference and analysis gene coexpression networks (GCNs)</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/chapters/02_network_comparison.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class='chapter-number'>2</span>  <span class='chapter-title'>Dealing with multiple data sets: consensus modules and module preservation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/chapters/03_network-based_data_integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>3</span>  <span class='chapter-title'>Network-based data integration for gene discovery</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/chapters/04_inference_and_analysis_of_GRNs.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>4</span>  <span class='chapter-title'>Inference and analysis of gene regulatory networks (GRNs)</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/chapters/appendix_01_data_acquisition.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Appendix: Data acquisition</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" ></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <div id="quarto-toc-target"></div>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">2</span>  <span class="chapter-title">Dealing with multiple data sets: consensus modules and module preservation</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>
<nav id="TOC" role="doc-toc">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#getting-to-know-the-example-data" id="toc-getting-to-know-the-example-data"><span class="header-section-number">2.1</span> Getting to know the example data</a></li>
  <li><a href="#data-preprocessing" id="toc-data-preprocessing"><span class="header-section-number">2.2</span> Data preprocessing</a></li>
  <li><a href="#identifying-and-analyzing-consensus-modules" id="toc-identifying-and-analyzing-consensus-modules"><span class="header-section-number">2.3</span> Identifying and analyzing consensus modules</a></li>
  <li><a href="#calculating-module-preservation-statistics" id="toc-calculating-module-preservation-statistics"><span class="header-section-number">2.4</span> Calculating module preservation statistics</a></li>
  <li><a href="#session-information" id="toc-session-information">Session information</a></li>
  <li><a href="#references" id="toc-references">References</a></li>
  </ul>
</nav>
<p>In this lesson, you will learn how to compare coexpression networks to identify preserved modules. At the end of the lesson, you will be able to:</p>
<ul>
<li>understand the concept of and identify consensus modules across data sets</li>
<li>associate consensus modules to traits</li>
<li>calculate module preservation statistics</li>
</ul>
<p>Let’s start by loading the packages we will use.</p>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>) <span class="co"># for reproducibility</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Load required packages</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BioNERO)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SummarizedExperiment)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(here)</span></code></pre></div>
</div>
<section id="getting-to-know-the-example-data" class="level2" data-number="2.1">
<h2 data-number="2.1"><span class="header-section-number">2.1</span> Getting to know the example data</h2>
<p>In this chapter, we will use gene expression data from two BioProjects:</p>
<ol type="1">
<li><p><strong>PRJNA800609</strong>: soybean pods infected with the fungus <em>Colletotrichum truncatum</em>. Original data generated by <span class="citation" data-cites="zhu2022transcriptomic">Zhu et al. (<a href="#ref-zhu2022transcriptomic" role="doc-biblioref">2022</a>)</span>.</p></li>
<li><p><strong>PRJNA574764</strong>: soybean roots infected with the oomycete <em>Phytophthora sojae</em>. Original data generated by <span class="citation" data-cites="de2020integrated">Ronne et al. (<a href="#ref-de2020integrated" role="doc-biblioref">2020</a>)</span>.</p></li>
</ol>
<p>Our goal here is to explore similarities and differences in expression profiles between these two data sets.</p>
<p>Data are available as <em>.rda</em> files in the <code>data/</code> directory of the <a href="https://github.com/almeidasilvaf/NASB">GitHub repo associated with this course</a>, and they were downloaded from <a href="https://soyatlas.venanciogroup.uenf.br/">The Soybean Expression Atlas v2</a> <span class="citation" data-cites="almeida2023soybean">(<a href="#ref-almeida2023soybean" role="doc-biblioref">Almeida-Silva, Pedrosa-Silva, and Venancio 2023</a>)</span> by searching by BioProject IDs. These <em>.rda</em> files contain <code>SummarizedExperiment</code> objects that store gene expression data and sample metadata.</p>
<p>Let’s load the data and explore them briefly.</p>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load expression data</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="fu">here</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;se_PRJNA800609.rda&quot;</span>))</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="fu">here</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;se_PRJNA574764.rda&quot;</span>))</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename object to a simpler name</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>exp1 <span class="ot">&lt;-</span> se_PRJNA800609</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>exp2 <span class="ot">&lt;-</span> se_PRJNA574764</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(se_PRJNA800609)</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(se_PRJNA574764)</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look at the object</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>exp1</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SummarizedExperiment 
dim: 31422 60 
metadata(0):
assays(1): &#39;&#39;
rownames(31422): Glyma.15G153300 Glyma.15G153400 ... Glyma.09G145600
  Glyma.09G145700
rowData names(0):
colnames(60): SAMN25263487 SAMN25263488 ... SAMN25263525 SAMN25263526
colData names(4): Part Cultivar Treatment Timepoint</code></pre>
</div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>exp2</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SummarizedExperiment 
dim: 32674 49 
metadata(0):
assays(1): &#39;&#39;
rownames(32674): Glyma.06G124400 Glyma.06G124500 ... Glyma.19G260900
  Glyma.19G261200
rowData names(0):
colnames(49): SAMN12868627 SAMN12868668 ... SAMN12868625 SAMN12868626
colData names(4): Part Cultivar Treatment Timepoint</code></pre>
</div>
<div class="sourceCode" id="cb6"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(exp1)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 60 rows and 4 columns
                    Part     Cultivar   Treatment   Timepoint
             &lt;character&gt;  &lt;character&gt; &lt;character&gt; &lt;character&gt;
SAMN25263487         pod         ZC-2     control          8h
SAMN25263488         pod         ZC-2     control          8h
SAMN25263507         pod         ZC-2    infected         48h
SAMN25263508         pod         ZC-2    infected         48h
SAMN25263527         pod Zhechun NO.3    infected         12h
...                  ...          ...         ...         ...
SAMN25263522         pod Zhechun NO.3     control         12h
SAMN25263523         pod Zhechun NO.3     control         12h
SAMN25263524         pod Zhechun NO.3     control         12h
SAMN25263525         pod Zhechun NO.3    infected         12h
SAMN25263526         pod Zhechun NO.3    infected         12h</code></pre>
</div>
<div class="sourceCode" id="cb8"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(exp2)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 49 rows and 4 columns
                    Part    Cultivar   Treatment   Timepoint
             &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;
SAMN12868627        root       Misty    infected       4 dpi
SAMN12868668        root   PI 449459    infected      14 dpi
SAMN12868669        root   PI 449459    infected      14 dpi
SAMN12868649        root   PI 449459     control       0 dpi
SAMN12868628        root       Misty    infected       7 dpi
...                  ...         ...         ...         ...
SAMN12868647        root   PI 449459     control       0 dpi
SAMN12868648        root   PI 449459     control       0 dpi
SAMN12868624        root       Misty     control       0 dpi
SAMN12868625        root       Misty    infected       4 dpi
SAMN12868626        root       Misty    infected       4 dpi</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Practice
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li>Explore the sample metadata of <code>exp1</code> and <code>exp2</code> and answer the questions below:</li>
</ol>
<ul>
<li>How many different cultivars are there?</li>
<li>What are the levels of the <code>Treatment</code> variable, and how many samples are there for each level?</li>
<li>How many samples are there for each timepoint?</li>
</ul>
<div class="callout callout-style-simple callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Show me the solutions
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Question 1: # of cultivars</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(exp1<span class="sc">$</span>Cultivar))</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2</code></pre>
</div>
<div class="sourceCode" id="cb12"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(exp2<span class="sc">$</span>Cultivar))</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2</code></pre>
</div>
<div class="sourceCode" id="cb14"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Question 2: levels of the `Treatment` variable</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(exp1<span class="sc">$</span>Treatment)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 control infected 
      30       30 </code></pre>
</div>
<div class="sourceCode" id="cb16"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(exp2<span class="sc">$</span>Treatment)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 control infected 
       6       43 </code></pre>
</div>
<div class="sourceCode" id="cb18"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Question 3: # of samples for each timepoint</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(exp1<span class="sc">$</span>Timepoint)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
12h 24h 48h  4h  8h 
 12  12  12  12  12 </code></pre>
</div>
<div class="sourceCode" id="cb20"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(exp2<span class="sc">$</span>Timepoint)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 0 dpi 14 dpi 21 dpi  4 dpi  7 dpi 
     6     12      8     12     11 </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="data-preprocessing" class="level2" data-number="2.2">
<h2 data-number="2.2"><span class="header-section-number">2.2</span> Data preprocessing</h2>
<p>Now, we will preprocess the two data sets using the same parameters with <code>exp_preprocess()</code>. In details, we will:</p>
<ol type="1">
<li>Keep only genes with median TPM &gt;=5.</li>
<li>Keep only the top 10k genes with the highest variances.</li>
</ol>
<div class="cell">
<div class="sourceCode" id="annotated-cell-6"><pre class="sourceCode r cell-code code-annotation-code"><code class="sourceCode r"><span id="annotated-cell-6-1"><a href="#annotated-cell-6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Store each expression data in a list, each data set in a list element</span></span>
<span id="annotated-cell-6-2"><a href="#annotated-cell-6-2" aria-hidden="true" tabindex="-1"></a>exp_list <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="annotated-cell-6-3"><a href="#annotated-cell-6-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">colletrotrichum_infection =</span> exp1,</span>
<span id="annotated-cell-6-4"><a href="#annotated-cell-6-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">phytophthora_infection =</span> exp2</span>
<span id="annotated-cell-6-5"><a href="#annotated-cell-6-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="annotated-cell-6-6"><a href="#annotated-cell-6-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-6-7"><a href="#annotated-cell-6-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop through the list and preprocess data</span></span>
<span id="annotated-cell-6-8"><a href="#annotated-cell-6-8" aria-hidden="true" tabindex="-1"></a>exp_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(</span>
<span id="annotated-cell-6-9"><a href="#annotated-cell-6-9" aria-hidden="true" tabindex="-1"></a>    exp_list,</span>
<span id="annotated-cell-6-10"><a href="#annotated-cell-6-10" aria-hidden="true" tabindex="-1"></a>    exp_preprocess,</span>
<span id="annotated-cell-6-11"><a href="#annotated-cell-6-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">min_exp =</span> <span class="dv">5</span>, <span class="at">variance_filter =</span> <span class="cn">TRUE</span>, <span class="at">n =</span> <span class="fl">1e4</span>, <span class="at">Zk_filtering =</span> <span class="cn">FALSE</span></span>
<span id="annotated-cell-6-12"><a href="#annotated-cell-6-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="annotated-cell-6-13"><a href="#annotated-cell-6-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-6-14"><a href="#annotated-cell-6-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Keep only genes that are shared between the two sets</span></span>
<span id="annotated-cell-6-15"><a href="#annotated-cell-6-15" aria-hidden="true" tabindex="-1"></a>shared <span class="ot">&lt;-</span> <span class="fu">intersect</span>(</span>
<span id="annotated-cell-6-16"><a href="#annotated-cell-6-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames</span>(exp_list<span class="sc">$</span>colletrotrichum_infection),</span>
<span id="annotated-cell-6-17"><a href="#annotated-cell-6-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames</span>(exp_list<span class="sc">$</span>phytophthora_infection)</span>
<span id="annotated-cell-6-18"><a href="#annotated-cell-6-18" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="annotated-cell-6-19"><a href="#annotated-cell-6-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-6-20"><a href="#annotated-cell-6-20" aria-hidden="true" tabindex="-1"></a>exp_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(exp_list, <span class="cf">function</span>(x) x[shared, ])</span></code></pre></div>
<div class="cell-annotation">
<dl>
<dt>Lines 2-5</dt>
<dd>
<span data-code-lines="2,3,4,5" data-code-cell="annotated-cell-6" data-code-annotation="1">Store each expression data set in a list element.</span>
</dd>
<dt>Lines 8-9</dt>
<dd>
<span data-code-lines="8,9" data-code-cell="annotated-cell-6" data-code-annotation="2">Loop through each element of the list <code>exp_list</code>, and</span>
</dd>
<dt>Line 10</dt>
<dd>
<span data-code-lines="10" data-code-cell="annotated-cell-6" data-code-annotation="3">execute the function <code>exp_preprocess</code>,</span>
</dd>
<dt>Line 11</dt>
<dd>
<span data-code-lines="11" data-code-cell="annotated-cell-6" data-code-annotation="4">using these parameters.</span>
</dd>
</dl>
</div>
</div>
<p>Now, we have a list of processed expression data. This list, with each element representing a different data set, is what we will use for all network comparison functions in the next sections.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Practice
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li><p>How many genes and samples are there in each processed data?</p></li>
<li><p>If we selected the top 10k genes with the highest variances, why do we not have 10k genes in each final set?</p></li>
</ol>
<div class="callout callout-style-simple callout-none no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Show me the solutions
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Q1</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sapply</span>(exp_list, nrow)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>colletrotrichum_infection    phytophthora_infection 
                     5530                      5530 </code></pre>
</div>
<div class="sourceCode" id="cb24"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sapply</span>(exp_list, ncol)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>colletrotrichum_infection    phytophthora_infection 
                       60                        49 </code></pre>
</div>
<div class="sourceCode" id="cb26"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; Q2: the genes with the highest variances are not the same in both sets.</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; Thus, when filtering the data to keep only shared genes, some genes are</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&#39; removed.</span></span></code></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="identifying-and-analyzing-consensus-modules" class="level2" data-number="2.3">
<h2 data-number="2.3"><span class="header-section-number">2.3</span> Identifying and analyzing consensus modules</h2>
<p>Consensus modules are coexpression modules present in different, independent data sets, and they can used to find robust modules across data sets that study the same (or similar) conditions.</p>
<p>To identify them, <strong>BioNERO</strong> infers a GCN for each data set and looks for groups of genes that are densely connected in all data sets. Thus, the workflow here will be very similar to what we did in the previous lesson. We will:</p>
<ol type="1">
<li>Identify the optimal <span class="math inline">\(\beta\)</span> power to which correlations will be raised (see previous chapter for more details on why this is done), but for each individual data set - <code>consensus_SFT_fit()</code></li>
<li>Infer GCNs and identify consensus modules - <code>consensus_modules()</code>.</li>
</ol>
<p>Let’s obtain the <span class="math inline">\(\beta\)</span> powers.</p>
<div class="cell">
<div class="sourceCode" id="cb27"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify the optimal beta power for each data set</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>sfts <span class="ot">&lt;-</span> <span class="fu">consensus_SFT_fit</span>(</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">exp_list =</span> exp_list,</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">setLabels =</span> <span class="fu">names</span>(exp_list),</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">cor_method =</span> <span class="st">&quot;pearson&quot;</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Power SFT.R.sq  slope truncated.R.sq mean.k. median.k. max.k.
1      5    0.924 -0.968          0.935   170.0    110.00    692
2      6    0.925 -1.050          0.936   127.0     75.40    589
3      7    0.923 -1.120          0.936    97.4     52.70    509
4      8    0.927 -1.160          0.944    76.5     37.10    445
5      9    0.928 -1.200          0.946    61.2     27.00    392
6     10    0.927 -1.230          0.947    49.7     19.80    349
7     11    0.929 -1.260          0.950    40.9     14.80    312
8     12    0.931 -1.290          0.954    34.0     11.10    281
9     13    0.940 -1.300          0.963    28.6      8.52    254
10    14    0.944 -1.320          0.968    24.2      6.55    231
11    15    0.948 -1.330          0.970    20.7      5.18    210
12    16    0.950 -1.340          0.973    17.8      4.03    192
13    17    0.954 -1.350          0.977    15.4      3.18    177
14    18    0.959 -1.350          0.982    13.4      2.51    163
15    19    0.960 -1.370          0.984    11.7      2.03    151
16    20    0.959 -1.380          0.985    10.3      1.64    140
   Power SFT.R.sq  slope truncated.R.sq mean.k. median.k. max.k.
1      5    0.866 -0.895          0.933  157.00    110.00    572
2      6    0.875 -1.000          0.942  116.00     71.10    493
3      7    0.874 -1.080          0.941   88.20     47.40    432
4      8    0.874 -1.140          0.936   68.90     32.40    382
5      9    0.844 -1.210          0.908   55.00     22.60    342
6     10    0.834 -1.260          0.893   44.70     16.10    308
7     11    0.828 -1.310          0.884   36.80     11.60    279
8     12    0.821 -1.340          0.875   30.70      8.48    255
9     13    0.785 -1.390          0.845   25.80      6.32    233
10    14    0.777 -1.420          0.843   22.00      4.77    215
11    15    0.791 -1.420          0.853   18.90      3.63    198
12    16    0.803 -1.420          0.863   16.30      2.79    184
13    17    0.813 -1.430          0.870   14.20      2.17    171
14    18    0.825 -1.430          0.882   12.40      1.69    159
15    19    0.800 -1.470          0.866   10.90      1.34    148
16    20    0.807 -1.460          0.874    9.66      1.06    139</code></pre>
</div>
<div class="sourceCode" id="cb29"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>sfts<span class="sc">$</span>power</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>colletrotrichum_infection    phytophthora_infection 
                        5                         5 </code></pre>
</div>
<div class="sourceCode" id="cb31"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>sfts<span class="sc">$</span>plot</span></code></pre></div>
<div class="cell-output-display">
<p><img src="02_network_comparison_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid" width="672" /></p>
</div>
</div>
<p>Next, let’s find consensus modules.</p>
<div class="cell">
<div class="sourceCode" id="cb32"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Find consensus modules</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>consensus <span class="ot">&lt;-</span> <span class="fu">consensus_modules</span>(</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>    exp_list, </span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">power =</span> sfts<span class="sc">$</span>power,</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">cor_method =</span> <span class="st">&quot;pearson&quot;</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>..connectivity..
..matrix multiplication (system BLAS)..
..normalization..
..done.
..connectivity..
..matrix multiplication (system BLAS)..
..normalization..
..done.
 ..done.
 multiSetMEs: Calculating module MEs.
   Working on set 1 ...
   Working on set 2 ...</code></pre>
</div>
<div class="sourceCode" id="cb34"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Taking a look at the consensus modules</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_dendro_and_colors</span>(consensus)</span></code></pre></div>
<div class="cell-output-display">
<p><img src="02_network_comparison_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="672" /></p>
</div>
<div class="sourceCode" id="cb35"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspecting the output</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(consensus)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] &quot;consMEs&quot;             &quot;exprSize&quot;            &quot;sampleInfo&quot;         
[4] &quot;genes_cmodules&quot;      &quot;dendro_plot_objects&quot;</code></pre>
</div>
</div>
<p>As you may have noticed, the output of <code>consensus_modules()</code> is very similar to the output of <code>exp2gcn()</code>. The output object is a list containing the following elements:</p>
<ol type="1">
<li><code>consMEs</code>: list with consensus module eigengenes.</li>
<li><code>exprSize</code>: list with number of data sets, and number of genes and samples for each set.</li>
<li><code>sampleInfo</code>: list of data frames with sample metadata.</li>
<li><code>genes_cmodules</code>: data frame with genes and their corresponding consensus modules.</li>
<li><code>dendro_plot_objects</code>: objects for plotting with <code>plot_dendro_and_colors()</code>.</li>
</ol>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Practice
</div>
</div>
<div class="callout-body-container callout-body">
<p>Explore the output of <code>consensus_modules()</code> and answer the following questions:</p>
<ol type="1">
<li>How many consensus modules were identified between the two data sets?</li>
<li>What are the largest and the smallest consensus modules?</li>
<li>What is the mean and median number of genes per consensus modules?</li>
</ol>
<div class="callout callout-style-simple callout-none no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Show me the solutions
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode" id="cb37"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Q1</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(consensus<span class="sc">$</span>genes_cmodules<span class="sc">$</span>Cons_modules))</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 30</code></pre>
</div>
<div class="sourceCode" id="cb39"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Q2</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sort</span>(<span class="fu">table</span>(consensus<span class="sc">$</span>genes_cmodules<span class="sc">$</span>Cons_modules))</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
         grey          blue         brown        yellow         green 
            8            34            38            54            57 
          red         black          pink        purple   greenyellow 
           61            66            68            75            77 
          tan        grey60     lightcyan    lightgreen   lightyellow 
           78            96            96            97            97 
      darkred darkturquoise      darkgrey        orange    darkorange 
          104           118           119           124           139 
        white       magenta   saddlebrown paleturquoise     royalblue 
          140           150           169           213           269 
       violet  midnightblue     darkgreen   yellowgreen          cyan 
          280           303           473           942           985 </code></pre>
</div>
<div class="sourceCode" id="cb41"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Q3</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(<span class="fu">table</span>(consensus<span class="sc">$</span>genes_cmodules<span class="sc">$</span>Cons_modules))</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 184.3333</code></pre>
</div>
<div class="sourceCode" id="cb43"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">median</span>(<span class="fu">table</span>(consensus<span class="sc">$</span>genes_cmodules<span class="sc">$</span>Cons_modules))</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 100.5</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<p>Next, you’d want to find correlations between consensus modules and traits of interest. Here, we will look for associations between consensus modules and the <code>Treatment</code> variable. Biologically speaking, we’re looking for shared transcriptional responses during infection with <em>Colletotrichum truncatum</em> and <em>Phytophthora sojae</em> (i.e., core immunity-related coexpression modules).</p>
<div class="cell">
<div class="sourceCode" id="cb45"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Correlate consensus modules to traits</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>consensus_trait <span class="ot">&lt;-</span> <span class="fu">consensus_trait_cor</span>(</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>    consensus,</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">metadata_cols =</span> <span class="st">&quot;Treatment&quot;</span></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Taking a look at the output</span></span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(consensus_trait)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    trait              ME         cor     pvalue     group
1 control         MEblack          NA         NA Treatment
2 control       MEdarkred  0.26015095 0.07092879 Treatment
3 control          MEblue -0.14934335 0.30747782 Treatment
4 control MEdarkturquoise  0.18536193 0.20341269 Treatment
5 control         MEbrown          NA         NA Treatment
6 control         MEgreen  0.05251631 0.69147472 Treatment</code></pre>
</div>
<div class="sourceCode" id="cb47"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot consensus module-trait correlations</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_module_trait_cor</span>(consensus_trait)</span></code></pre></div>
<div class="cell-output-display">
<p><img src="02_network_comparison_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid" width="480" /></p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Practice
</div>
</div>
<div class="callout-body-container callout-body">
<p>Explore the output of <code>consensus_trait_cor()</code> and answer the questions below:</p>
<ol type="1">
<li><p>Which consensus module has the highest positive correlation to the <em>infected</em> status of the <code>Treatment</code> variable?</p></li>
<li><p>Which consensus module has the highest negative correlation to the <em>infected</em> status of the <code>Treatment</code> variable?</p></li>
<li><p>(Advanced) Based on your biological knowledge, what gene functions would you expect to find in the modules you found in questions 1 and 2?</p></li>
</ol>
<div class="callout callout-style-simple callout-none no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Show me the solutions
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode" id="cb48"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Q1</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>cor_infected <span class="ot">&lt;-</span> consensus_trait[consensus_trait<span class="sc">$</span>trait <span class="sc">==</span> <span class="st">&quot;infected&quot;</span>, ]</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>cor_infected[<span class="fu">which.max</span>(cor_infected<span class="sc">$</span>cor), ]</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      trait    ME       cor      pvalue     group
59 infected MEred 0.4029435 0.003765978 Treatment</code></pre>
</div>
<div class="sourceCode" id="cb50"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Q2</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>cor_infected[<span class="fu">which.min</span>(cor_infected<span class="sc">$</span>cor), ]</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      trait             ME        cor      pvalue     group
49 infected MEmidnightblue -0.3966709 0.004420417 Treatment</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Challenge
</div>
</div>
<div class="callout-body-container callout-body">
<p>Use the function <code>module_enrichment()</code> to perform a functional enrichment analysis for each consensus module. Then, try to interpret the results in light of the consensus module-trait associations you found previously.</p>
<p>Tip: to load the functional annotation data, use the following:</p>
<div class="cell">
<div class="sourceCode" id="cb52"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="fu">here</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;gma_annotation.rda&quot;</span>))</span></code></pre></div>
</div>
<div class="callout callout-style-simple callout-none no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Show me the solutions
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode" id="cb53"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load functional annotation data</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="fu">here</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;gma_annotation.rda&quot;</span>))</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Define background</span></span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>background <span class="ot">&lt;-</span> <span class="fu">rownames</span>(exp_list<span class="sc">$</span>colletrotrichum_infection)</span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform the enrichment analysis</span></span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a>sim_net <span class="ot">&lt;-</span> consensus</span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sim_net)[<span class="dv">4</span>] <span class="ot">&lt;-</span> <span class="st">&quot;genes_and_modules&quot;</span></span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sim_net<span class="sc">$</span>genes_and_modules) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Genes&quot;</span>, <span class="st">&quot;Modules&quot;</span>)</span>
<span id="cb53-11"><a href="#cb53-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-12"><a href="#cb53-12" aria-hidden="true" tabindex="-1"></a>enrich_mapman <span class="ot">&lt;-</span> <span class="fu">module_enrichment</span>(sim_net, background, gma_annotation<span class="sc">$</span>MapMan)</span>
<span id="cb53-13"><a href="#cb53-13" aria-hidden="true" tabindex="-1"></a>enrich_interpro <span class="ot">&lt;-</span> <span class="fu">module_enrichment</span>(sim_net, background, gma_annotation<span class="sc">$</span>InterPro)</span></code></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="calculating-module-preservation-statistics" class="level2" data-number="2.4">
<h2 data-number="2.4"><span class="header-section-number">2.4</span> Calculating module preservation statistics</h2>
<p>When we infer consensus modules across data sets, we only consider shared modules, but we have no information on which modules are <strong>not shared</strong> between different data sets.</p>
<p>If you want to have a more detailed picture of which modules are preserved and which are not, you’d need to infer a separate network for each data set, and then calculate module preservation statistics between the networks.</p>
<p>Here, we will demonstrate how to do that using the same data set from the previous section. Since, we already have the processed data, we will proceed to GCN inference using the functions <code>SFT_fit()</code> and <code>exp2gcn()</code>, as we saw in Chapter 1.</p>
<div class="cell">
<div class="sourceCode" id="cb54"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get optimal beta power for each data set</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>powers <span class="ot">&lt;-</span> <span class="fu">lapply</span>(exp_list, SFT_fit, <span class="at">cor_method =</span> <span class="st">&quot;pearson&quot;</span>)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Power SFT.R.sq  slope truncated.R.sq mean.k. median.k. max.k.
1      3    0.234 -0.774         0.0153  1130.0     997.0   1810
2      4    0.372 -0.725         0.2440   805.0     699.0   1510
3      5    0.530 -0.728         0.4860   602.0     509.0   1290
4      6    0.654 -0.757         0.6480   466.0     382.0   1130
5      7    0.767 -0.816         0.7850   372.0     295.0   1000
6      8    0.840 -0.872         0.8600   303.0     233.0    898
7      9    0.884 -0.923         0.9010   251.0     186.0    813
8     10    0.901 -0.968         0.9190   211.0     151.0    742
9     11    0.912 -1.010         0.9280   179.0     123.0    681
10    12    0.920 -1.050         0.9380   154.0     101.0    628
11    13    0.921 -1.080         0.9390   134.0      84.5    582
12    14    0.914 -1.110         0.9310   117.0      70.4    541
13    15    0.919 -1.140         0.9390   102.0      59.5    504
14    16    0.921 -1.160         0.9410    90.5      50.5    471
15    17    0.923 -1.180         0.9420    80.4      42.9    441
16    18    0.925 -1.200         0.9460    71.7      36.3    415
17    19    0.927 -1.220         0.9480    64.3      30.9    390
18    20    0.931 -1.240         0.9520    57.8      26.7    368
   Power SFT.R.sq   slope truncated.R.sq mean.k. median.k. max.k.
1      3  0.66100  4.4900          0.947  1140.0    1140.0   1510
2      4  0.44400  2.0600          0.883   803.0     797.0   1230
3      5  0.16500  0.6540          0.685   594.0     580.0   1050
4      6  0.00224 -0.0438          0.585   456.0     432.0    921
5      7  0.28100 -0.4140          0.709   359.0     330.0    818
6      8  0.64700 -0.6550          0.878   290.0     255.0    735
7      9  0.78900 -0.7960          0.926   238.0     199.0    669
8     10  0.82900 -0.8870          0.937   198.0     158.0    613
9     11  0.84100 -0.9530          0.933   168.0     126.0    566
10    12  0.85100 -1.0100          0.937   143.0     103.0    524
11    13  0.86000 -1.0500          0.943   123.0      83.2    488
12    14  0.86000 -1.0900          0.941   107.0      68.1    456
13    15  0.86100 -1.1200          0.940    93.7      56.2    428
14    16  0.85800 -1.1600          0.934    82.5      46.5    402
15    17  0.86800 -1.1700          0.939    73.0      38.7    379
16    18  0.85500 -1.2100          0.925    65.0      32.4    359
17    19  0.83400 -1.2400          0.905    58.2      27.2    340
18    20  0.83600 -1.2600          0.905    52.3      23.1    322</code></pre>
</div>
<div class="sourceCode" id="cb56"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Infer GCN for each data set</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>gcns <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">seq_along</span>(powers), <span class="cf">function</span>(n) {</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>    gcn <span class="ot">&lt;-</span> <span class="fu">exp2gcn</span>(</span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>        exp_list[[n]], </span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">SFTpower =</span> powers[[n]]<span class="sc">$</span>power, </span>
<span id="cb56-7"><a href="#cb56-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">cor_method =</span> <span class="st">&quot;pearson&quot;</span></span>
<span id="cb56-8"><a href="#cb56-8" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb56-9"><a href="#cb56-9" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb56-10"><a href="#cb56-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(gcn)</span>
<span id="cb56-11"><a href="#cb56-11" aria-hidden="true" tabindex="-1"></a>})</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>..connectivity..
..matrix multiplication (system BLAS)..
..normalization..
..done.
..connectivity..
..matrix multiplication (system BLAS)..
..normalization..
..done.</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Practice
</div>
</div>
<div class="callout-body-container callout-body">
<p>How many modules are there in each network?</p>
<div class="callout callout-style-simple callout-none no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-11-contents" aria-controls="callout-11" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Show me the solutions
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-11" class="callout-11-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode" id="cb58"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sapply</span>(gcns, <span class="cf">function</span>(x) <span class="fu">length</span>(<span class="fu">unique</span>(x<span class="sc">$</span>genes_and_modules<span class="sc">$</span>Modules)))</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 12 17</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<p>Next, we can calculate module preservation statistics using the permutation-based approach implemented in <em><a href="https://CRAN.R-project.org/package=NetRep">NetRep</a></em>.</p>
<div class="cell">
<div class="sourceCode" id="cb60"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate module preservation statistics</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>pres_netrep <span class="ot">&lt;-</span> <span class="fu">module_preservation</span>(</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>    exp_list, </span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">ref_net =</span> gcns[[<span class="dv">1</span>]], </span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">test_net =</span> gcns[[<span class="dv">2</span>]], </span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">algorithm =</span> <span class="st">&quot;netrep&quot;</span></span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[2023-08-24 09:56:26 CEST] Validating user input...
[2023-08-24 09:56:26 CEST]   Checking matrices for problems...
[2023-08-24 09:56:28 CEST] Input ok!
[2023-08-24 09:56:28 CEST] Calculating preservation of network subsets from
                           dataset &quot;colletrotrichum_infection&quot; in dataset
                           &quot;phytophthora_infection&quot;.
[2023-08-24 09:56:28 CEST]   Pre-computing network properties in dataset
                             &quot;colletrotrichum_infection&quot;...
[2023-08-24 09:56:29 CEST]   Calculating observed test statistics...
[2023-08-24 09:56:29 CEST]   Generating null distributions from 1000
                             permutations using 1 thread...


    0% completed.
    1% completed.
    2% completed.
    2% completed.
    3% completed.
    4% completed.
    5% completed.
    5% completed.
    6% completed.
    7% completed.
    8% completed.
    9% completed.
    9% completed.
   10% completed.
   11% completed.
   12% completed.
   13% completed.
   13% completed.
   14% completed.
   15% completed.
   16% completed.
   16% completed.
   17% completed.
   18% completed.
   19% completed.
   20% completed.
   20% completed.
   21% completed.
   22% completed.
   23% completed.
   24% completed.
   24% completed.
   25% completed.
   26% completed.
   27% completed.
   28% completed.
   28% completed.
   29% completed.
   30% completed.
   31% completed.
   31% completed.
   32% completed.
   33% completed.
   34% completed.
   35% completed.
   35% completed.
   36% completed.
   37% completed.
   38% completed.
   39% completed.
   39% completed.
   40% completed.
   41% completed.
   42% completed.
   42% completed.
   43% completed.
   44% completed.
   45% completed.
   46% completed.
   46% completed.
   47% completed.
   48% completed.
   49% completed.
   50% completed.
   50% completed.
   51% completed.
   52% completed.
   53% completed.
   54% completed.
   54% completed.
   55% completed.
   56% completed.
   57% completed.
   57% completed.
   58% completed.
   59% completed.
   60% completed.
   61% completed.
   61% completed.
   62% completed.
   63% completed.
   64% completed.
   64% completed.
   65% completed.
   66% completed.
   67% completed.
   68% completed.
   68% completed.
   69% completed.
   70% completed.
   71% completed.
   72% completed.
   72% completed.
   73% completed.
   74% completed.
   75% completed.
   75% completed.
   76% completed.
   77% completed.
   78% completed.
   79% completed.
   79% completed.
   80% completed.
   81% completed.
   82% completed.
   83% completed.
   83% completed.
   84% completed.
   85% completed.
   86% completed.
   87% completed.
   87% completed.
   88% completed.
   89% completed.
   90% completed.
   90% completed.
   91% completed.
   92% completed.
   93% completed.
   94% completed.
   94% completed.
   95% completed.
   96% completed.
   97% completed.
   98% completed.
   98% completed.
   99% completed.
  100% completed.
  100% completed.

[2023-08-24 09:58:37 CEST]   Calculating P-values...
[2023-08-24 09:58:38 CEST]   Collating results...
[2023-08-24 09:58:39 CEST] Done!</code></pre>
</div>
<div class="sourceCode" id="cb62"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Taking a look at the P-values for preservation statistics for each module</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pres_netrep<span class="sc">$</span>p.values)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           avg.weight   coherence     cor.cor  cor.degree cor.contrib
black     0.000999001 0.000999001 0.000999001 0.000999001 0.000999001
blue      0.142857143 0.262737263 0.000999001 0.000999001 0.096903097
cyan      0.994005994 0.999000999 0.000999001 0.000999001 0.000999001
darkgreen 0.976023976 1.000000000 0.000999001 0.001998002 0.994005994
green     0.002997003 1.000000000 0.000999001 0.000999001 0.000999001
grey      0.986013986 1.000000000 0.001998002 0.176823177 0.138861139
              avg.cor avg.contrib
black     0.000999001 0.000999001
blue      0.000999001 0.188811189
cyan      0.000999001 0.000999001
darkgreen 0.000999001 0.100899101
green     0.000999001 0.000999001
grey      0.005994006 0.544455544</code></pre>
</div>
</div>
<p>Note that, to calculate module preservation statistics, you always need to choose a <strong>reference network</strong> and a <strong>test network</strong>. Thus, the function <code>module_preservation()</code> will return which modules of the reference network that are preserved in the test network.</p>
<p>Careful readers will also notice that this is another major difference between identifying consensus modules and calculating module preservation statistics: one can identify consensus modules across any number of data sets, but module preservation statistics can only be calculated in a pairwise manner.</p>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Practice
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li><p>By default, <strong>BioNERO</strong> considers modules to be preserved if all preservation statistics (<span class="math inline">\(N=7\)</span>) are significant (<span class="math inline">\(P&lt;0.05\)</span>). Suppose you want to be less stringent and consider modules as preserved if 5 or more statistics are significant. Would the number of preserved modules change? Verify that.</p></li>
<li><p>Use the function <code>module_enrichment()</code> to find enriched functions in all modules of the reference network. Then, explore the enriched functions of preserved modules (if any).</p></li>
</ol>
<div class="callout callout-style-simple callout-none no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-13-contents" aria-controls="callout-13" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Show me the solutions
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-13" class="callout-13-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode" id="cb64"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Q1</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="do">## Extract significant statistics for each module</span></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>sig_stats <span class="ot">&lt;-</span> <span class="fu">apply</span>(pres_netrep<span class="sc">$</span>p.values, <span class="dv">1</span>, <span class="cf">function</span>(x) x[x <span class="sc">&lt;</span> <span class="fl">0.05</span>])</span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>sig_stats</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$black
 avg.weight   coherence     cor.cor  cor.degree cor.contrib     avg.cor 
0.000999001 0.000999001 0.000999001 0.000999001 0.000999001 0.000999001 
avg.contrib 
0.000999001 

$blue
    cor.cor  cor.degree     avg.cor 
0.000999001 0.000999001 0.000999001 

$cyan
    cor.cor  cor.degree cor.contrib     avg.cor avg.contrib 
0.000999001 0.000999001 0.000999001 0.000999001 0.000999001 

$darkgreen
    cor.cor  cor.degree     avg.cor 
0.000999001 0.001998002 0.000999001 

$green
 avg.weight     cor.cor  cor.degree cor.contrib     avg.cor avg.contrib 
0.002997003 0.000999001 0.000999001 0.000999001 0.000999001 0.000999001 

$grey
    cor.cor     avg.cor 
0.001998002 0.005994006 

$grey60
 avg.weight   coherence     cor.cor  cor.degree cor.contrib     avg.cor 
0.000999001 0.000999001 0.000999001 0.000999001 0.000999001 0.000999001 
avg.contrib 
0.000999001 

$lightcyan
 avg.weight   coherence     cor.cor  cor.degree     avg.cor avg.contrib 
0.000999001 0.000999001 0.000999001 0.000999001 0.000999001 0.000999001 

$red
    cor.cor  cor.degree     avg.cor 
0.001998002 0.045954046 0.010989011 

$royalblue
    cor.cor  cor.degree     avg.cor avg.contrib 
0.000999001 0.000999001 0.000999001 0.048951049 

$salmon
    cor.cor  cor.degree cor.contrib     avg.cor avg.contrib 
0.000999001 0.000999001 0.000999001 0.000999001 0.000999001 

$skyblue
 avg.weight   coherence     cor.cor  cor.degree     avg.cor avg.contrib 
0.000999001 0.000999001 0.000999001 0.042957043 0.000999001 0.000999001 </code></pre>
</div>
<div class="sourceCode" id="cb66"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Keep only modules with 5+ significant stats</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>sig_stats[<span class="fu">lengths</span>(sig_stats) <span class="sc">&gt;=</span> <span class="dv">5</span>]</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$black
 avg.weight   coherence     cor.cor  cor.degree cor.contrib     avg.cor 
0.000999001 0.000999001 0.000999001 0.000999001 0.000999001 0.000999001 
avg.contrib 
0.000999001 

$cyan
    cor.cor  cor.degree cor.contrib     avg.cor avg.contrib 
0.000999001 0.000999001 0.000999001 0.000999001 0.000999001 

$green
 avg.weight     cor.cor  cor.degree cor.contrib     avg.cor avg.contrib 
0.002997003 0.000999001 0.000999001 0.000999001 0.000999001 0.000999001 

$grey60
 avg.weight   coherence     cor.cor  cor.degree cor.contrib     avg.cor 
0.000999001 0.000999001 0.000999001 0.000999001 0.000999001 0.000999001 
avg.contrib 
0.000999001 

$lightcyan
 avg.weight   coherence     cor.cor  cor.degree     avg.cor avg.contrib 
0.000999001 0.000999001 0.000999001 0.000999001 0.000999001 0.000999001 

$salmon
    cor.cor  cor.degree cor.contrib     avg.cor avg.contrib 
0.000999001 0.000999001 0.000999001 0.000999001 0.000999001 

$skyblue
 avg.weight   coherence     cor.cor  cor.degree     avg.cor avg.contrib 
0.000999001 0.000999001 0.000999001 0.042957043 0.000999001 0.000999001 </code></pre>
</div>
<div class="sourceCode" id="cb68"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Q2</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>background <span class="ot">&lt;-</span> <span class="fu">rownames</span>(exp_list[[<span class="dv">1</span>]])</span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>enrich_preserved <span class="ot">&lt;-</span> <span class="fu">module_enrichment</span>(</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>    gcns[[<span class="dv">1</span>]],</span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">background_genes =</span> background,</span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">annotation =</span> gma_annotation<span class="sc">$</span>MapMan</span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="session-information" class="level2 unnumbered">
<h2 class="unnumbered">Session information</h2>
<p>This chapter was created under the following conditions:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>─ Session info ───────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.0 (2023-04-21)
 os       Ubuntu 20.04.5 LTS
 system   x86_64, linux-gnu
 ui       X11
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       Europe/Brussels
 date     2023-08-24
 pandoc   3.1.1 @ /usr/lib/rstudio/resources/app/bin/quarto/bin/tools/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────────
 package              * version   date (UTC) lib source
 abind                  1.4-5     2016-07-21 [1] CRAN (R 4.3.0)
 annotate               1.78.0    2023-04-25 [1] Bioconductor
 AnnotationDbi          1.62.0    2023-04-25 [1] Bioconductor
 backports              1.4.1     2021-12-13 [1] CRAN (R 4.3.0)
 base64enc              0.1-3     2015-07-28 [1] CRAN (R 4.3.0)
 Biobase              * 2.60.0    2023-04-25 [1] Bioconductor
 BiocGenerics         * 0.46.0    2023-04-25 [1] Bioconductor
 BiocManager            1.30.21.1 2023-07-18 [1] CRAN (R 4.3.0)
 BiocParallel           1.34.0    2023-04-25 [1] Bioconductor
 BiocStyle              2.29.1    2023-08-04 [1] Github (Bioconductor/BiocStyle@7c0e093)
 BioNERO              * 1.9.7     2023-08-23 [1] Bioconductor
 Biostrings             2.68.0    2023-04-25 [1] Bioconductor
 bit                    4.0.5     2022-11-15 [1] CRAN (R 4.3.0)
 bit64                  4.0.5     2020-08-30 [1] CRAN (R 4.3.0)
 bitops                 1.0-7     2021-04-24 [1] CRAN (R 4.3.0)
 blob                   1.2.4     2023-03-17 [1] CRAN (R 4.3.0)
 cachem                 1.0.8     2023-05-01 [1] CRAN (R 4.3.0)
 Cairo                  1.6-0     2022-07-05 [1] CRAN (R 4.3.0)
 checkmate              2.2.0     2023-04-27 [1] CRAN (R 4.3.0)
 circlize               0.4.15    2022-05-10 [1] CRAN (R 4.3.0)
 cli                    3.6.1     2023-03-23 [1] CRAN (R 4.3.0)
 clue                   0.3-64    2023-01-31 [1] CRAN (R 4.3.0)
 cluster                2.1.4     2022-08-22 [4] CRAN (R 4.2.1)
 coda                   0.19-4    2020-09-30 [1] CRAN (R 4.3.0)
 codetools              0.2-19    2023-02-01 [4] CRAN (R 4.2.2)
 colorspace             2.1-0     2023-01-23 [1] CRAN (R 4.3.0)
 ComplexHeatmap         2.16.0    2023-04-25 [1] Bioconductor
 crayon                 1.5.2     2022-09-29 [1] CRAN (R 4.3.0)
 data.table             1.14.8    2023-02-17 [1] CRAN (R 4.3.0)
 DBI                    1.1.3     2022-06-18 [1] CRAN (R 4.3.0)
 DelayedArray           0.26.1    2023-05-01 [1] Bioconductor
 digest                 0.6.33    2023-07-07 [1] CRAN (R 4.3.0)
 doParallel             1.0.17    2022-02-07 [1] CRAN (R 4.3.0)
 dplyr                * 1.1.2     2023-04-20 [1] CRAN (R 4.3.0)
 dynamicTreeCut         1.63-1    2016-03-11 [1] CRAN (R 4.3.0)
 edgeR                  3.42.0    2023-04-25 [1] Bioconductor
 evaluate               0.21      2023-05-05 [1] CRAN (R 4.3.0)
 fansi                  1.0.4     2023-01-22 [1] CRAN (R 4.3.0)
 farver                 2.1.1     2022-07-06 [1] CRAN (R 4.3.0)
 fastcluster            1.2.3     2021-05-24 [1] CRAN (R 4.3.0)
 fastmap                1.1.1     2023-02-24 [1] CRAN (R 4.3.0)
 forcats              * 1.0.0     2023-01-29 [1] CRAN (R 4.3.0)
 foreach                1.5.2     2022-02-02 [1] CRAN (R 4.3.0)
 foreign                0.8-82    2022-01-13 [4] CRAN (R 4.1.2)
 Formula                1.2-5     2023-02-24 [1] CRAN (R 4.3.0)
 genefilter             1.82.0    2023-04-25 [1] Bioconductor
 generics               0.1.3     2022-07-05 [1] CRAN (R 4.3.0)
 GENIE3                 1.22.0    2023-04-25 [1] Bioconductor
 GenomeInfoDb         * 1.36.0    2023-04-25 [1] Bioconductor
 GenomeInfoDbData       1.2.10    2023-04-28 [1] Bioconductor
 GenomicRanges        * 1.52.0    2023-04-25 [1] Bioconductor
 GetoptLong             1.0.5     2020-12-15 [1] CRAN (R 4.3.0)
 ggdendro               0.1.23    2022-02-16 [1] CRAN (R 4.3.0)
 ggnetwork              0.5.12    2023-03-06 [1] CRAN (R 4.3.0)
 ggplot2              * 3.4.1     2023-02-10 [1] CRAN (R 4.3.0)
 ggrepel                0.9.3     2023-02-03 [1] CRAN (R 4.3.0)
 GlobalOptions          0.1.2     2020-06-10 [1] CRAN (R 4.3.0)
 glue                   1.6.2     2022-02-24 [1] CRAN (R 4.3.0)
 GO.db                  3.17.0    2023-05-02 [1] Bioconductor
 gridExtra              2.3       2017-09-09 [1] CRAN (R 4.3.0)
 gtable                 0.3.3     2023-03-21 [1] CRAN (R 4.3.0)
 here                 * 1.0.1     2020-12-13 [1] CRAN (R 4.3.0)
 Hmisc                  5.0-1     2023-03-08 [1] CRAN (R 4.3.0)
 hms                    1.1.3     2023-03-21 [1] CRAN (R 4.3.0)
 htmlTable              2.4.1     2022-07-07 [1] CRAN (R 4.3.0)
 htmltools              0.5.5     2023-03-23 [1] CRAN (R 4.3.0)
 htmlwidgets            1.6.2     2023-03-17 [1] CRAN (R 4.3.0)
 httr                   1.4.5     2023-02-24 [1] CRAN (R 4.3.0)
 igraph                 1.4.2     2023-04-07 [1] CRAN (R 4.3.0)
 impute                 1.74.0    2023-04-25 [1] Bioconductor
 intergraph             2.0-2     2016-12-05 [1] CRAN (R 4.3.0)
 IRanges              * 2.34.0    2023-04-25 [1] Bioconductor
 iterators              1.0.14    2022-02-05 [1] CRAN (R 4.3.0)
 jsonlite               1.8.7     2023-06-29 [1] CRAN (R 4.3.0)
 KEGGREST               1.40.0    2023-04-25 [1] Bioconductor
 knitr                  1.43      2023-05-25 [1] CRAN (R 4.3.0)
 labeling               0.4.2     2020-10-20 [1] CRAN (R 4.3.0)
 lattice                0.20-45   2021-09-22 [4] CRAN (R 4.2.0)
 lifecycle              1.0.3     2022-10-07 [1] CRAN (R 4.3.0)
 limma                  3.56.0    2023-04-25 [1] Bioconductor
 locfit                 1.5-9.7   2023-01-02 [1] CRAN (R 4.3.0)
 lubridate            * 1.9.2     2023-02-10 [1] CRAN (R 4.3.0)
 magick                 2.7.4     2023-03-09 [1] CRAN (R 4.3.0)
 magrittr               2.0.3     2022-03-30 [1] CRAN (R 4.3.0)
 MASS                   7.3-58.2  2023-01-23 [4] CRAN (R 4.2.2)
 Matrix                 1.5-1     2022-09-13 [4] CRAN (R 4.2.1)
 MatrixGenerics       * 1.12.2    2023-06-09 [1] Bioconductor
 matrixStats          * 1.0.0     2023-06-02 [1] CRAN (R 4.3.0)
 memoise                2.0.1     2021-11-26 [1] CRAN (R 4.3.0)
 mgcv                   1.8-41    2022-10-21 [4] CRAN (R 4.2.1)
 minet                  3.58.0    2023-04-25 [1] Bioconductor
 munsell                0.5.0     2018-06-12 [1] CRAN (R 4.3.0)
 NetRep                 1.2.6     2023-01-06 [1] CRAN (R 4.3.0)
 network                1.18.1    2023-01-24 [1] CRAN (R 4.3.0)
 nlme                   3.1-162   2023-01-31 [4] CRAN (R 4.2.2)
 nnet                   7.3-18    2022-09-28 [4] CRAN (R 4.2.1)
 patchwork              1.1.2     2022-08-19 [1] CRAN (R 4.3.0)
 pillar                 1.9.0     2023-03-22 [1] CRAN (R 4.3.0)
 pkgconfig              2.0.3     2019-09-22 [1] CRAN (R 4.3.0)
 plyr                   1.8.8     2022-11-11 [1] CRAN (R 4.3.0)
 png                    0.1-8     2022-11-29 [1] CRAN (R 4.3.0)
 preprocessCore         1.62.0    2023-04-25 [1] Bioconductor
 purrr                * 1.0.1     2023-01-10 [1] CRAN (R 4.3.0)
 R6                     2.5.1     2021-08-19 [1] CRAN (R 4.3.0)
 RColorBrewer           1.1-3     2022-04-03 [1] CRAN (R 4.3.0)
 Rcpp                   1.0.10    2023-01-22 [1] CRAN (R 4.3.0)
 RCurl                  1.98-1.12 2023-03-27 [1] CRAN (R 4.3.0)
 readr                * 2.1.4     2023-02-10 [1] CRAN (R 4.3.0)
 reshape2               1.4.4     2020-04-09 [1] CRAN (R 4.3.0)
 RhpcBLASctl            0.23-42   2023-02-11 [1] CRAN (R 4.3.0)
 rjson                  0.2.21    2022-01-09 [1] CRAN (R 4.3.0)
 rlang                  1.1.1     2023-04-28 [1] CRAN (R 4.3.0)
 rmarkdown              2.23      2023-07-01 [1] CRAN (R 4.3.0)
 rpart                  4.1.19    2022-10-21 [4] CRAN (R 4.2.1)
 rprojroot              2.0.3     2022-04-02 [1] CRAN (R 4.3.0)
 RSQLite                2.3.1     2023-04-03 [1] CRAN (R 4.3.0)
 rstudioapi             0.14      2022-08-22 [1] CRAN (R 4.3.0)
 S4Arrays               1.0.1     2023-05-01 [1] Bioconductor
 S4Vectors            * 0.38.0    2023-04-25 [1] Bioconductor
 scales                 1.2.1     2022-08-20 [1] CRAN (R 4.3.0)
 sessioninfo            1.2.2     2021-12-06 [1] CRAN (R 4.3.0)
 shape                  1.4.6     2021-05-19 [1] CRAN (R 4.3.0)
 statmod                1.5.0     2023-01-06 [1] CRAN (R 4.3.0)
 statnet.common         4.8.0     2023-01-24 [1] CRAN (R 4.3.0)
 stringi                1.7.12    2023-01-11 [1] CRAN (R 4.3.0)
 stringr              * 1.5.0     2022-12-02 [1] CRAN (R 4.3.0)
 SummarizedExperiment * 1.30.1    2023-05-01 [1] Bioconductor
 survival               3.5-3     2023-02-12 [4] CRAN (R 4.2.2)
 sva                    3.48.0    2023-04-25 [1] Bioconductor
 tibble               * 3.2.1     2023-03-20 [1] CRAN (R 4.3.0)
 tidyr                * 1.3.0     2023-01-24 [1] CRAN (R 4.3.0)
 tidyselect             1.2.0     2022-10-10 [1] CRAN (R 4.3.0)
 tidyverse            * 2.0.0     2023-02-22 [1] CRAN (R 4.3.0)
 timechange             0.2.0     2023-01-11 [1] CRAN (R 4.3.0)
 tzdb                   0.3.0     2022-03-28 [1] CRAN (R 4.3.0)
 utf8                   1.2.3     2023-01-31 [1] CRAN (R 4.3.0)
 vctrs                  0.6.3     2023-06-14 [1] CRAN (R 4.3.0)
 WGCNA                  1.72-1    2023-01-18 [1] CRAN (R 4.3.0)
 withr                  2.5.0     2022-03-03 [1] CRAN (R 4.3.0)
 xfun                   0.39      2023-04-20 [1] CRAN (R 4.3.0)
 XML                    3.99-0.14 2023-03-19 [1] CRAN (R 4.3.0)
 xtable                 1.8-4     2019-04-21 [1] CRAN (R 4.3.0)
 XVector                0.40.0    2023-04-25 [1] Bioconductor
 yaml                   2.3.7     2023-01-23 [1] CRAN (R 4.3.0)
 zlibbioc               1.46.0    2023-04-25 [1] Bioconductor

 [1] /home/faalm/R/x86_64-pc-linux-gnu-library/4.3
 [2] /usr/local/lib/R/site-library
 [3] /usr/lib/R/site-library
 [4] /usr/lib/R/library

──────────────────────────────────────────────────────────────────────────────</code></pre>
</div>
</div>
</section>
<section id="references" class="level2 unnumbered">
<h2 class="unnumbered">References</h2>
<div id="quarto-navigation-envelope" class="hidden">
<p><span class="hidden" data-render-id="quarto-int-sidebar-title">Network Analysis in Systems Biology with R/Bioconductor</span> <span class="hidden" data-render-id="quarto-int-navbar-title">Network Analysis in Systems Biology with R/Bioconductor</span> <span class="hidden" data-render-id="quarto-int-next"><span class="chapter-number">3</span>  <span class="chapter-title">Network-based data integration for gene discovery</span></span> <span class="hidden" data-render-id="quarto-int-prev"><span class="chapter-number">1</span>  <span class="chapter-title">Inference and analysis gene coexpression networks (GCNs)</span></span> <span class="hidden" data-render-id="quarto-int-sidebar:/index.html">Preface</span> <span class="hidden" data-render-id="quarto-int-sidebar:/chapters/00_setup.html">Setup</span> <span class="hidden" data-render-id="quarto-int-sidebar:/chapters/01_inference_and_analysis_of_GCNs.html"><span class="chapter-number">1</span>  <span class="chapter-title">Inference and analysis gene coexpression networks (GCNs)</span></span> <span class="hidden" data-render-id="quarto-int-sidebar:/chapters/02_network_comparison.html"><span class="chapter-number">2</span>  <span class="chapter-title">Dealing with multiple data sets: consensus modules and module preservation</span></span> <span class="hidden" data-render-id="quarto-int-sidebar:/chapters/03_network-based_data_integration.html"><span class="chapter-number">3</span>  <span class="chapter-title">Network-based data integration for gene discovery</span></span> <span class="hidden" data-render-id="quarto-int-sidebar:/chapters/04_inference_and_analysis_of_GRNs.html"><span class="chapter-number">4</span>  <span class="chapter-title">Inference and analysis of gene regulatory networks (GRNs)</span></span> <span class="hidden" data-render-id="quarto-int-sidebar:/chapters/appendix_01_data_acquisition.html">Appendix: Data acquisition</span> <span class="hidden" data-render-id="quarto-breadcrumbs-ecd29dd0b7b9b53310ca7a0e98f18f5f"><span class="chapter-number">2</span>  <span class="chapter-title">Dealing with multiple data sets: consensus modules and module preservation</span></span></p>
</div>
<div id="quarto-meta-markdown" class="hidden">
<p><span class="hidden" data-render-id="quarto-metatitle">Network Analysis in Systems Biology with R/Bioconductor - <span class="chapter-number">2</span>  <span class="chapter-title">Dealing with multiple data sets: consensus modules and module preservation</span></span> <span class="hidden" data-render-id="quarto-twittercardtitle">Network Analysis in Systems Biology with R/Bioconductor - <span class="chapter-number">2</span>  <span class="chapter-title">Dealing with multiple data sets: consensus modules and module preservation</span></span> <span class="hidden" data-render-id="quarto-ogcardtitle">Network Analysis in Systems Biology with R/Bioconductor - <span class="chapter-number">2</span>  <span class="chapter-title">Dealing with multiple data sets: consensus modules and module preservation</span></span> <span class="hidden" data-render-id="quarto-metasitename">Network Analysis in Systems Biology with R/Bioconductor</span> <span class="hidden" data-render-id="quarto-twittercarddesc"></span> <span class="hidden" data-render-id="quarto-ogcardddesc"></span></p>
</div>
<div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-almeida2023soybean" class="csl-entry" role="listitem">
Almeida-Silva, Fabricio, Francisnei Pedrosa-Silva, and Thiago M Venancio. 2023. <span>“The Soybean Expression Atlas V2: A Comprehensive Database of over 5000 RNA-Seq Samples.”</span> <em>bioRxiv</em>, 2023–04.
</div>
<div id="ref-de2020integrated" class="csl-entry" role="listitem">
Ronne, Maxime de, Caroline Labbé, Amandine Lebreton, Humira Sonah, Rupesh Deshmukh, Martine Jean, François Belzile, Louise O’Donoughue, and Richard Bélanger. 2020. <span>“Integrated QTL Mapping, Gene Expression and Nucleotide Variation Analyses to Investigate Complex Quantitative Traits: A Case Study with the Soybean–Phytophthora Sojae Interaction.”</span> <em>Plant Biotechnology Journal</em> 18 (7): 1492.
</div>
<div id="ref-zhu2022transcriptomic" class="csl-entry" role="listitem">
Zhu, Longming, Qinghua Yang, Xiaomin Yu, Xujun Fu, Hangxia Jin, and Fengjie Yuan. 2022. <span>“Transcriptomic and Metabolomic Analyses Reveal a Potential Mechanism to Improve Soybean Resistance to Anthracnose.”</span> <em>Frontiers in Plant Science</em> 13: 850829.
</div>
</div>
</section>

</main> <!-- /main -->
<script id = "quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a  href="/chapters/01_inference_and_analysis_of_GCNs.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class='chapter-number'>1</span>  <span class='chapter-title'>Inference and analysis gene coexpression networks (GCNs)</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a  href="/chapters/03_network-based_data_integration.html" class="pagination-link">
        <span class="nav-page-text"><span class='chapter-number'>3</span>  <span class='chapter-title'>Network-based data integration for gene discovery</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->

</body>

</html>